Revised syntax to avoid the need for an import

djspiewak opened the pull request:
This now works:

```scala
async[IO] {
  ioa.await + iob.await
}
```

Unfortunately, so does this!

```scala
val ioa: IO[A] = ???
async[OptionT[IO, *]] {
  ioa.await
}
```

So uh… yeah. I tried fiddling around with making `await` take an implicit parameter that evidences the `F`, and then materializing that parameter in the `async` macro expansion, but unfortunately macros are expanded inside-out rather than outside-in. In fact, we actually *leverage* this exact trick now with the `_await` expansion.

Also detecting this lexically at compile time might be really hard because of nesting: `async[IO](await(async[OptionT[IO, *]](ioa.await).value))` (where `ioa: IO[A]`). Ultimately, `markForAsyncTransform` probably needs to be richer so that we can ensure that it only transforms `await`s which are bound to our scope, and perhaps gives us the ability to partially apply and re-do some typechecking. Maybe there's some trick we can use though. At the very least, taking an `<:<` might allow us to detect things at runtime and avoid hanging `Dispatcher`.

djspiewak commented:
@retronym You may be interested in some of the limitations we're hitting here. Not having a direct way of threading a context from the `async` to the `await` is really the problem. Additionally, note the shenanigans we have to go through in order to get `body.await` instead of `await(body)`.

Baccata commented on core/shared/src/main/scala/cats/effect/cps/AsyncAwait.scala:
```diff
@@ -88,49 +85,81 @@ object AsyncAwaitDsl {
-              ${c.prefix}._AsyncInstance.async_[${c.prefix}._AsyncContext[AnyRef]](cb => new $name(dispatcher, cb).start())
+          ${F}.flatten {
+            _root_.cats.effect.std.Dispatcher[$effect].use { dispatcher =>
+              ${F}.async_[$name#FF[AnyRef]](cb => new $name(dispatcher, cb).start())
```
This is somewhat infuriating, but couldn't figure for the life of me how to apply an argument to the `effect` type value (which represents the polymorphic F). So instead, added a type member to `AsyncAwaitStateMachine` so that I'm able to reference it using a type projection ... happy for it to be changed if you know how to.

djspiewak replied:
LOL wow that's annoying. I'm not that worried about it though. We can also iterate on it later.

Baccata commented on core/shared/src/main/scala/cats/effect/cps/AsyncAwait.scala:
```diff
@@ -35,45 +35,42 @@ import cats.syntax.all._
  * an "await" block.
  */
-class AsyncAwaitDsl[F[_]](implicit F: Async[F]) {
+object dsl {
```
Should we move the contents to a `cps` package object, so that people just have to import `cats.effect.cps._` ?

djspiewak replied:
Yep! Also should just be a regular object, not a package object, since I don't think we actually need to nest any types.

retronym commented:
Yeah, this is an annoying part of the embedding into Scala.

Kotlin has a better story as the `await` method can injects an "implicit receiver" (the DSL containing `await`) into the scope.

In Scala 2.x, that would be too boilerplatey:

```scala
case class Future[T](t: T)
def async[T](f: AwaitDSL => T): Future[T] = ???
trait AwaitDSL { def await[U](f: Future[U]): U = ??? }

object Test {
  val x, y = Future(42)

  async {
    dsl => import dsl._ // BOILERPLATE

    await(x) + await(y)
  }
}
```

We can reduce the boilerplate a little by using the lambda parameter as an implicit:

```scala
case class Future[T](t: T)
trait CanAwait
def async[T](f: CanAwait => T): Future[T] = ???
def await[U](f: Future[U])(implicit ev: CanAwait): U = ???

object Test {
  val x, y = Future(42)
  async { implicit ~ => 
    await(x) + await(y)
  }
}
```

The remnant boilerplate is still more than users would like (and take some explaining!), and you need to be careful to choose the same name for the implicit parameter in nested contexts to avoid "ambiguous implicit" errors.

Scala 3 contextual functions runs with this pattern by fixing those two problems.

To tie this to scala-async, you could implement the DSL above as macros that expand to underlying `_async` / `_await` methods.

djspiewak commented:
@retronym I think that part of the problem is that we have different sorts of `await`s. As an example:

```scala
val ioa: IO[A] = ???
val optIOA: OptionT[IO, A] = ??

async[IO] {
  ioa.await    // valid!
  optIOA.await   // invalid!

  val opt2 = async[OptionT[IO, *]] {
    optIOA.await  // valid
    ioa.await   //invalid
  }

  opt2.value.await  // valid
}
```

You see the problem. :-) Just a simple `CanAwait` isn't sufficient, but we also can't necessarily do something like `CanAwait[IO]` because then the inner `ioa.await` would be allowed.

There's also a somewhat larger problem here, which is that the `await` macro would be expanded (to `_await`) prior to the expansion of the `async` macro, meaning that it becomes an implicit error after the inner expansion.

Also re-flagging the pain around `foo.await` vs `await(foo)` as something that's probably easier to address. :-)

djspiewak commented on core/shared/src/test/scala/cats/effect/cps/AsyncAwaitSpec.scala:
```diff
@@ -201,23 +194,57 @@ class AsyncAwaitSpec extends Specification with CatsEffect {
+      val tc = typecheck("async[OptionTIO](IO(1).await)").result
+      tc must beLike {
+        case TypecheckError(message) =>
+          message must contain("Expected await to be called on")
```
Minor nit, but compilation errors are idiomatically spelled in all lower-case.

retronym commented:
Yeah, there just isn't language support for automatically obscuring a value from an enclosing scope, especially once you give them different types (e.g. making `CanAwait` path-dependent and/or parametric.)

Kotlin introduced [`@DslMarker`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/-dsl-marker/) to solve this problem. The motivating use case was a HTML DSL, to exclude:

```kotlin
table { 
  tr {
    div {
      tr // compile error, the innermost implicit implicit receiver for `div` is part of the the same DSL as the enclosing one for `table` / `tr`.
```

Brainstorming a little: I wonder if we could simulate this with a macro. Let's say `await` was a macro that internally performed a custom implicit search for `CanAwait`, rather than accepting the implicit parameter. It could search for `CanAwait[_]` as well as the `CanAwait[SomeF]` and error if they aren't the same instance.

I'm pretty sure that you can make `foo.await` work, at least by making it a macro that expands to `await(foo)`.

Baccata commented:
@retronym all the problems discussed between yourself and @djspiewak are already worked-around/solved in this very PR 😄 

> I'm pretty sure that you can make foo.await work, at least by making it a macro that expands to await(foo).

As for the contextual binding between async and await, we've implemented a [piece of logic](https://github.com/typelevel/cats-effect-cps/pull/3/files#diff-92546d3a8bdfb79716caee74c7245ff288d0629462a17db618e7ed44ad16f3cdR67-R74) to pre-process the tree that gets passed to `markForAsyncTransform` to ensure that the `F` types in `await` calls match the one in the `async` call, and abort the macro expansion if they don't. It's a bit ad-hoc but seems to work well enough.

djspiewak commented:
@retronym Expanding on the following a bit:

> As for the contextual binding between async and await, we've implemented a piece of logic to pre-process the tree that gets passed to markForAsyncTransform, in order to ensure that the F types in await calls match the one in the async call, and abort the macro expansion if they don't. It's a bit ad-hoc but seems to work well enough.

I think a better design here would be if `markForAsyncTransform` itself allowed us to pre-process the `await` calls *it* identifies as it goes through the transformation. The problem is that, in the limit (and especially as the async transform evolves and becomes more advanced), the traversal we implemented is basically replicating the same logic which has to exist within `markForAsyncTransform` itself. Having some sort of pre-processing mechanism applied during transform would resolve that duplication and potentially open the door to a more robust transform in the future.
